java7/lib/src/main/java/com/truelayer/signing/Signer.java

```diff
@@ -0,0 +1,221 @@
+     */
+    public Signer header(String name, String value) {
+        if (name == null || value == null)
+            throw new IllegalArgumentException("header name and value must not be null");
+        this.headers.put(new HeaderName(name), value);
+        return this;
+    }
+
+    /**
+     * Add a Map of headers.
+     * Warning: Only a single value per header name is supported.
+     *
+     * @param headers  - must not be null
+     * @return the Signer instance
+     * @throws IllegalArgumentException if the provided params are null
+     */
+    public Signer headers(Map<String, String> headers) {
+        if (headers == null)
+            throw new IllegalArgumentException("headers must not be null");
+        for (Map.Entry<String, String> entry : headers.entrySet()) {
+            this.headers.put(new HeaderName(entry.getKey()), entry.getValue());
+        }
+        return this;
+    }
+
+    /**
+     * Start building a request Tl-Signature header value using private key
+     * RFC 7468 PEM-encoded data and the key's kid.
+     *
+     * @param kid           key identifier of the private key - must not be null
+     * @param privateKeyPem the privateKey RFC 7468 PEM-encoded - must not be null
+     * @return the Signer instance
+     * @throws KeyException             if the provided key is invalid
+     * @throws IllegalArgumentException if the provided params are null
+     */
+    public static Signer from(String kid, byte[] privateKeyPem) {
+        if (kid == null || privateKeyPem == null)
+            throw new IllegalArgumentException("kid and privateKey must not be null");
+
+        ECPrivateKey privateKey;
+        try {
+            privateKey = ECKey.parseFromPEMEncodedObjects(new String(privateKeyPem)).toECKey().toECPrivateKey();
+        } catch (Exception e) {
+            throw new KeyException("Exception thrown while parsing private key", e);
+        }
+
+        return new Signer(kid, privateKey);
+    }
+
+    /**
+     * Start building a request Tl-Signature header value using private key
+     * RFC 7468 PEM-encoded data and the key's kid.
+     *
+     * @param kid           key identifier of the private key - must not be null
+     * @param privateKeyPem the privateKey RFC 7468 PEM-encoded - must not be null
+     * @return the Signer instance
+     * @throws KeyException             if the provided key is invalid
+     * @throws IllegalArgumentException if the provided params are null
+     */
+    public static Signer from(String kid, String privateKeyPem) {
+        if (kid == null || privateKeyPem == null)
+            throw new IllegalArgumentException("kid and privateKey must not be null");
+
+        ECPrivateKey privateKey;
+        try {
+            privateKey = ECKey.parseFromPEMEncodedObjects(privateKeyPem).toECKey().toECPrivateKey();
+        } catch (Exception e) {
+            throw new KeyException("Exception thrown while parsing private key", e);
```

Comment:
Is it possible to send a more meaningful exception message? Something that can help to understand the actual issue?

Reply:
There is the `e` which details more what happened. However, I will change this anyway, and just throw the message in the exception itself